zio-http/src/main/scala/zio/http/netty/AsyncBodyReader.scala

```diff
@@ -38,11 +38,22 @@ abstract class AsyncBodyReader(implicit trace: Trace) extends SimpleChannelInbou
     this.synchronized {
       state match {
         case State.Buffering =>
-          state = State.Direct(callback)
-          buffer.result().foreach { case (chunk, isLast) =>
-            callback(chunk, isLast)
+          val result: Chunk[(Chunk[Byte], Boolean)] = buffer.result()
+          val readingDone: Boolean                  = result.lastOption match {
+            case None              => false
+            case Some((_, isLast)) => isLast
           }
-          ctx.read()
+
+          if (ctx.channel.isOpen || readingDone) {
```

Comment:
I'm not sure if it's expected or not, but in some cases this fix won't be triggered.
It is possible that reading will be done before we try to connect to a body, thus we will still be able to read from a closed channel.
Even with your changes, the following code sometimes succeeds
```scala
package example

import zio._
import zio.http._

object SimpleClient extends ZIOAppDefault {
  def run: RIO[Scope, Any]               = program.provideSome[Scope](Client.default)
  def program: RIO[Scope & Client, Unit] = for {
    client <- ZIO.service[Client]
    request = client.request(Request.get("https://jsonplaceholder.typicode.com/todos/1"))
    res <- ZIO.scoped(request)
    _   <- res.body.asString.debug
  } yield ()
}
```
I say "sometimes" because sometimes it fails with mysterious 
```
timestamp=2023-09-23T14:16:20.575404Z level=ERROR thread=#zio-fiber-1 message="" cause="Exception in thread "zio-fiber-19,4" io.netty.channel.AbstractChannel$AnnotatedNoRouteToHostException: No route to host: jsonplaceholder.typicode.com/[2606:4700:e4:0:0:0:ac40:a206]:443
```
Which is not related to your changes (looks like another bug, lol)

Reply:
I'm aware of the flakiness you suggested. But the check I added there is not a feature so much as a requirement. Like, is it *required* to keep the channel open while we read it (ie Is it the case that in the future, our code will never close the channel before the AsyncBodyReader.connect is called?)

If that's the case, I'm all for failing whenever the channel is closed. If not, then the check should stay there.